libraries/chain/controller.cpp

```diff
@@ -1318,6 +1359,16 @@ struct controller_impl {
 
                db.commit( (*bitr)->block_num() );
                root_id = (*bitr)->id();
+
+               if constexpr (std::is_same_v<block_state_legacy_ptr, std::decay_t<decltype(*bitr)>>) {
+                  if ((*bitr)->header.contains_header_extension(instant_finality_extension::extension_id())) {
+                     savanna_transistion_required = true;
+                     // Do not advance irreversible past IF Genesis Block
+                     break;
+                  }
+               } else if ((*bitr)->block->is_proper_svnn_block()) {
+                  fork_db.switch_to_savanna();
```

Comment:
Where do we use `fork_db_l`?

Reply:
The apply will use `fork_db_l` until `switch_to_savanna()` is called. So it seems that if the above `if constexpr` was false, allowing us to execute line 1370, it would mean that `switch_to_savanna()` has already been called.

Reply:
`fork_db_l` is used until the `fork_db.switch_from_legacy(new_root)` in `transition_to_savanna()`. When `in_use_t::both` then `fork_db.apply()` uses `fork_db_s`.

Reply:
> When `in_use_t::both` then `fork_db.apply()` uses `fork_db_s`.

[does it?](https://github.com/AntelopeIO/leap/blob/f576797b9bec31b5af02be9126dc20e3536b07dd/libraries/chain/include/eosio/chain/fork_database.hpp#L183-L198)

Reply:
Yes, it only uses `fork_db_l` when `in_use.load() == in_use_t::legacy`

Reply:
> Yes, it only uses `fork_db_l` when `in_use.load() == in_use_t::legacy`

You're right, I went off on a tangent.

Got it, this code is activated when we are in `both` mode to switch to `savanna` only mode. Thanks!